  \documentclass{beamer}
\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{listings}
\lstset{frame=single,backgroundcolor=\color{lightgray}}
\usetheme{Warsaw}

\newcommand{\remoteimage}[3]{
\IfFileExists{#1}{}{\immediate\write18{curl -o "#1" "#2"}}
\begin{center}
\includegraphics[#3]{#1}
\end{center}
}

\newcommand{\graphviz}[3]{
\IfFileExists{#1}{}{\immediate\write18{echo #2 | dot -o"#1.png" -Tpng}}
\begin{center}
\includegraphics[#3]{#1.png}
\end{center}
}

\newcommand{\centeredtitle}[1]{
\begin{center}
    \Huge{\bf{#1}}
\end{center}
}

\newcommand{\hugeslide}[1]{
\begin{frame}
\centeredtitle{#1}
\end{frame}
}



\title{Advanced NCBI.\\The Entrez API\\\href{https://github.com/lindenb/courses}{https://github.com/lindenb/courses}}
\author{Pierre Lindenbaum\\\href{https://twitter.com/yokofakun}{@yokofakun}\\ \href{mailto:plindenbaum@yahoo.fr}{pierre.lindenbaum@univ-nantes.fr}\\ \url{http://plindenbaum.blogspot.com} }\institute{Institut du Thorax. Nantes. France}

\begin{document}



\begin{frame} 
\titlepage
\end{frame}


\begin{frame}[fragile]
NCBI ? What about EBI, ENSEMBL, ...
\end{frame}

\begin{frame}
\remoteimage{jeter01.png}{http://www.ncbi.nlm.nih.gov/corehtml/pmc/pmcgifs/bookshelf/thumbs/th-helpeutils-lrg.png}{}
\end{frame}


\begin{frame}
What will be covered today? :
\begin{itemize}
\item File formats...
\item EInfo, GQuery, ESearch , Esummary, EFetch..
\item processing XML answer with XSLT: HTML, SVG, R...
\item generating a java parser for dbSNP.
\item NCBI EBot
\item using standalone BLAST
\end{itemize}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% TOOLS
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
\begin{frame}[fragile]
\frametitle{CURL}
\begin{lstlisting}[language=bash]
curl "http://en.wikipedia.org/wiki/Main_page"
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{XML}
\remoteimage{jeter02.png}{https://upload.wikimedia.org/wikipedia/commons/thumb/6/68/XML.svg/500px-XML.svg.png}{scale=0.3}
\end{frame}


\begin{frame}[fragile]
\frametitle{XSLT}
\remoteimage{jeter03.png}{https://upload.wikimedia.org/wikipedia/commons/thumb/e/e6/XSLT_en.svg/231px-XSLT_en.svg.png}{scale=0.4}
\end{frame}


\begin{frame}[fragile]
\frametitle{XSLTPROC}
\begin{lstlisting}[language=bash]
xsltproc stylesheet.xsl file.xml > result.xml
\end{lstlisting}
\end{frame}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% FORMAT
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\hugeslide{Formats}

\begin{frame}[fragile]
\frametitle{Formats}
\framesubtitle{Genbank}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=25&rettype=gb}
\begin{lstlisting}[basicstyle=\tiny,breaklines=false]
LOCUS       X53813                   422 bp    DNA     linear   MAM 22-JUN-1992
DEFINITION  Blue Whale heavy satellite DNA.
ACCESSION   X53813 X17460
VERSION     X53813.1  GI:25
KEYWORDS    satellite DNA.
SOURCE      Balaenoptera musculus (Blue whale)
  ORGANISM  Balaenoptera musculus
            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
            Mammalia; Eutheria; Laurasiatheria; Cetartiodactyla; Cetacea;
            Mysticeti; Balaenopteridae; Balaenoptera.
REFERENCE   1  (bases 1 to 422)
  AUTHORS   Arnason,U. and Widegren,B.
  TITLE     Composition and chromosomal localization of cetacean highly
            repetitive DNA with special reference to the blue whale,
            Balaenoptera musculus
  JOURNAL   Chromosoma 98 (5), 323-329 (1989)
   PUBMED   2612291
COMMENT     See also <X52700-2> for 1,760 bp common cetacean component clones
            and <X52703-6>,<X53811-4> for the 422 bp heavy satellite clones.
FEATURES             Location/Qualifiers
     source          1..422
                     /organism="Balaenoptera musculus"
                     /mol_type="genomic DNA"
                     /db_xref="taxon:9771"
                     /clone="7"
     misc_feature    1..422
                     /note="heavy satellite DNA"
ORIGIN      
        1 tagttattca acctatccca ctctctagat accccttagc acgtaaagga atattatttg
       61 ggggtccagc catggagaat agtttagaca ctaggatgag ataaggaaca cacccattct
      121 aaagaaatca cattaggatt ctctttttaa gctgttcctt aaaacactag agtcttagaa
      181 atctattgga ggcagaagca gtcaagggta gcctagggtt agggttaggc ttagggttag
      241 ggttagggta cggcttaggg tactgtttcg gggaggggtt caggtacggc gtagggtatg
      301 ggttagggtt agggttaggg ttagtgttag ggttagggct cggtttaggg tacgggttag
      361 gattagggta cgtgttaggg ttagggtagg gcttagggtt agggtacgtg ttagggttag
      421 gg
//
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=25}
\frametitle{Formats}
\framesubtitle{ASN.1}
\begin{lstlisting}[basicstyle=\tiny,breaklines=false]
Seq-entry ::= seq {
  id {
    embl {
      accession "X53813" ,
      version 1 } ,
    gi 25 } ,
  descr {
    title "Blue Whale heavy satellite DNA" ,
    source {
      org {
        taxname "Balaenoptera musculus" ,
        common "Blue whale" ,
        db {
          {
            db "taxon" ,
            tag
              id 9771 } } ,
        orgname {
          name
            binomial {
              genus "Balaenoptera" ,
              species "musculus" } ,
          lineage "Eukaryota; Metazoa; Chordata; Craniata; Vertebrata;
 Euteleostomi; Mammalia; Eutheria; Laurasiatheria; Cetartiodactyla; Cetacea;
 Mysticeti; Balaenopteridae; Balaenoptera" ,
          gcode 1 ,
          mgcode 2 ,
          div "MAM" } } ,
      subtype {
        {
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Formats}
\framesubtitle{ASN.1 (schema)}
\url{http://www.ncbi.nlm.nih.gov/data_specs/asn/insdseq.asn}
\begin{lstlisting}[basicstyle=\tiny,breaklines=false]
INSDSeq ::= SEQUENCE {
    locus VisibleString ,
    length INTEGER ,
    strandedness VisibleString OPTIONAL ,
    moltype VisibleString ,
    topology VisibleString OPTIONAL ,
    division VisibleString ,
    update-date VisibleString ,
    create-date VisibleString OPTIONAL ,
    update-release VisibleString OPTIONAL ,
    create-release VisibleString OPTIONAL ,
    definition VisibleString ,
    primary-accession VisibleString OPTIONAL ,
    entry-version VisibleString OPTIONAL ,
    accession-version VisibleString OPTIONAL ,
    other-seqids SEQUENCE OF INSDSeqid OPTIONAL ,
    secondary-accessions SEQUENCE OF INSDSecondary-accn OPTIONAL,
    project VisibleString OPTIONAL ,
    keywords SEQUENCE OF INSDKeyword OPTIONAL ,
    segment VisibleString OPTIONAL ,
    source VisibleString OPTIONAL ,
    organism VisibleString OPTIONAL ,
    taxonomy VisibleString OPTIONAL ,
    references SEQUENCE OF INSDReference OPTIONAL ,
    comment VisibleString OPTIONAL ,
    comment-set SEQUENCE OF INSDComment OPTIONAL ,
    struc-comments SEQUENCE OF INSDStrucComment OPTIONAL ,
    primary VisibleString OPTIONAL ,
    source-db VisibleString OPTIONAL ,
    database-reference VisibleString OPTIONAL ,
    feature-table SEQUENCE OF INSDFeature OPTIONAL ,
    feature-set SEQUENCE OF INSDFeatureSet OPTIONAL ,
    sequence VisibleString OPTIONAL ,  -- Optional for contig, wgs, etc.
    contig VisibleString OPTIONAL ,
    alt-seq SEQUENCE OF INSDAltSeqData OPTIONAL
}
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{Formats}
\framesubtitle{ASN.1 (tools)}
\begin{center}
{\bf DATATOOL}\\
Generate {\bf C++ }data storage classes based on ASN.1
{\bf serialization streams}.\\
Convert data between ASN.1, XML and JSON formats.\\
\end{center}
\end{frame}

\begin{frame}[fragile]
\frametitle{Formats}
\framesubtitle{XML}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=25&retmode=xml}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<?xml version="1.0"?>
 <!DOCTYPE GBSet PUBLIC "-//NCBI//NCBI GBSeq/EN" "http://www.ncbi.nlm.nih.gov/dtd/NCBI_GBSeq.dtd">
 <GBSet>
<GBSeq>
  <GBSeq_locus>X53813</GBSeq_locus>
  <GBSeq_length>422</GBSeq_length>
  <GBSeq_strandedness>double</GBSeq_strandedness>
  <GBSeq_moltype>DNA</GBSeq_moltype>
  <GBSeq_topology>linear</GBSeq_topology>
  <GBSeq_division>MAM</GBSeq_division>
  <GBSeq_update-date>22-JUN-1992</GBSeq_update-date>
  <GBSeq_create-date>13-JUL-1990</GBSeq_create-date>
  <GBSeq_definition>Blue Whale heavy satellite DNA</GBSeq_definition>
  <GBSeq_primary-accession>X53813</GBSeq_primary-accession>
  <GBSeq_accession-version>X53813.1</GBSeq_accession-version>
  <GBSeq_other-seqids>
    <GBSeqid>emb|X53813.1|</GBSeqid>
    <GBSeqid>gi|25</GBSeqid>
  </GBSeq_other-seqids>
  <GBSeq_secondary-accessions>
    <GBSecondary-accn>X17460</GBSecondary-accn>
  </GBSeq_secondary-accessions>
  <GBSeq_keywords>
    <GBKeyword>satellite DNA</GBKeyword>
  </GBSeq_keywords>
  <GBSeq_source>Balaenoptera musculus (Blue whale)</GBSeq_source>
  <GBSeq_organism>Balaenoptera musculus</GBSeq_organism>
  <GBSeq_taxonomy>Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia; Eutheria; Laurasiatheria; Cetartiod
actyla; Cetacea; Mysticeti; Balaenopteridae; Balaenoptera</GBSeq_taxonomy>
  <GBSeq_references>
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Formats}
\framesubtitle{XML (DTD)}
\url{http://www.ncbi.nlm.nih.gov/dtd/NCBI_GBSeq.mod.dtd}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<!ELEMENT GBSeq (
        GBSeq_locus, 
        GBSeq_length, 
        GBSeq_strandedness?, 
        GBSeq_moltype, 
        GBSeq_topology?, 
        GBSeq_division, 
        GBSeq_update-date, 
        GBSeq_create-date?, 
        GBSeq_update-release?, 
        GBSeq_create-release?, 
        GBSeq_definition, 
        GBSeq_primary-accession?, 
        GBSeq_entry-version?, 
        GBSeq_accession-version?, 
        GBSeq_other-seqids?, 
        GBSeq_secondary-accessions?, 
        GBSeq_project?, 
        GBSeq_keywords?, 
        GBSeq_segment?, 
        GBSeq_source?, 
        GBSeq_organism?, 
        GBSeq_taxonomy?, 
        GBSeq_references?, 
        GBSeq_comment?, 
        GBSeq_comment-set?, 
        GBSeq_struc-comments?, 
        (...)
\end{lstlisting}
\end{frame}


\hugeslide{E-Utilities}

\hugeslide{GI}


\begin{frame}
\frametitle{E-Utilities}
\begin{center}
Set of seven server-side programs that provide a stable interface to the search, retrieval, and linking functions of the Entrez system, using a fixed URL syntax.\\
The output provided by the E-Utilities is in XML format\\
\end{center}
\end{frame}


\begin{frame}
\frametitle{Entrez Direct}
\begin{center}
\url{http://www.ncbi.nlm.nih.gov/books/NBK179288/}
"Entrez Direct (EDirect) is an advanced method for accessing the NCBI's set of interconnected databases (publication, sequence, structure, gene, variation, expression, etc.) from a UNIX terminal window. Functions take search terms from command-line arguments. Individual operations are combined to build multi-step queries. Record retrieval and formatting normally complete the process."
\end{center}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% EINFO
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\hugeslide{EInfo}

\begin{frame}[fragile]
\frametitle{EInfo}
\begin{center}
Provides a {\bf list of the names of all valid Entrez databases}.\\
Provides statistics for a single database, including lists of indexing fields and available link names.
\end{center}
\end{frame}

\begin{frame}[fragile]
\frametitle{EInfo}
Base URL:
\small
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi}
\end{frame}



\begin{frame}[fragile]
\frametitle{EInfo}
\framesubtitle{XML Ouput}

\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi}

\small
\begin{lstlisting}[language=xml]
<!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd">
<eInfoResult>
  <DbList>
    <DbName>pubmed</DbName>
    <DbName>protein</DbName>
    <DbName>nuccore</DbName>
    <DbName>nucleotide</DbName>
    <DbName>nucgss</DbName>
    <DbName>nucest</DbName>
    <DbName>structure</DbName>
    <DbName>genome</DbName>
    <DbName>assembly</DbName>
    <DbName>gcassembly</DbName>
    <DbName>genomeprj</DbName>
    <DbName>bioproject</DbName>
    <DbName>biosample</DbName>
    <DbName>biosystems</DbName>
    <DbName>blastdbinfo</DbName>
    <DbName>books</DbName>
    <DbName>cdd</DbName>
    <DbName>clinvar</DbName>
 (...)
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{EInfo}
Return statistics for a given Entrez database:\\
\small
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=DbName}
\end{frame}


\begin{frame}[fragile]
\frametitle{EInfo}
\framesubtitle{Statistics for Pubmed}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=pubmed}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<?xml version="1.0"?>
<!DOCTYPE eInfoResult PUBLIC "-//NLM//DTD eInfoResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd">
<eInfoResult>
  <DbInfo>
    <DbName>pubmed</DbName>
    <MenuName>PubMed</MenuName>
    <Description>PubMed bibliographic record</Description>
    <DbBuild>Build130805-2117m.4</DbBuild>
    <Count>22974581</Count>
    <LastUpdate>2013/08/06 08:33</LastUpdate>
    <FieldList>
      (...)
      <Field>
        <Name>UID</Name>
        <FullName>UID</FullName>
        <Description>Unique number assigned to publication</Description>
        <TermCount>0</TermCount>
        <IsDate>N</IsDate>
        <IsNumerical>Y</IsNumerical>
        <SingleToken>Y</SingleToken>
        <Hierarchy>N</Hierarchy>
        <IsHidden>Y</IsHidden>
      </Field>
      <Field>
(...)
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{EInfo}
\framesubtitle{Statistics for Pubmed}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/einfo.fcgi?db=pubmed}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
(...)
      <Link>
        <Name>pubmed_taxonomy_entrez</Name>
        <Menu>Taxonomy via GenBank</Menu>
        <Description>Taxonomy records associated with the current articles through taxonomic information on related molecular database records (Nucleotide, Protein, Gene, SNP, Structu
re).</Description>
        <DbTo>taxonomy</DbTo>
      </Link>
      <Link>
        <Name>pubmed_unigene</Name>
        <Menu>UniGene Links</Menu>
        <Description>UniGene clusters of expressed sequences that are associated with the current articles through references on the clustered sequence records and related Gene record
s.</Description>
        <DbTo>unigene</DbTo>
      </Link>
      <Link>
        <Name>pubmed_unists</Name>
        <Menu>UniSTS Links</Menu>
        <Description>Genetic, physical, and sequence mapping reagents in the UniSTS database associated with the current articles through references on sequence tagged site (STS) subm
issions as well as automated searching of PubMed abstracts and full-text PubMed Central articles for marker names.</Description>
        <DbTo>unists</DbTo>
      </Link>
    </LinkList>
  </DbInfo>
</eInfoResult>
\end{lstlisting}
\end{frame}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% GQuery
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\hugeslide{GQuery}


\begin{frame}[fragile]
\frametitle{GQuery}
Provides the number of records retrieved in all Entrez databases by a single text query.
\end{frame}



\begin{frame}[fragile]
\frametitle{GQuery}
\framesubtitle{Example}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl  "http://eutils.ncbi.nlm.nih.gov/gquery?term=tyrannosaurus%20rex&retmode=xml"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=true,escapechar=\%]
<Result>
    <Term>tyrannosaurus rex</Term>
    <eGQueryResult>
        <ResultItem><DbName>pubmed</DbName><MenuName/><Count>41</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>pmc</DbName><MenuName/><Count>160</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>mesh</DbName><MenuName/><Count>15</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>books</DbName><MenuName/><Count>179</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>pubmedhealth</DbName><MenuName/><Count>21</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>omim</DbName><MenuName/><Count>10</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>omia</DbName><MenuName/><Count>0</Count><Status>Term or Database is not found</Status></ResultItem>
        <ResultItem><DbName>ncbisearch</DbName><MenuName/><Count>1</Count><Status>Ok</Status></ResultItem>
        <ResultItem><DbName>nuccore</DbName><MenuName/><Count>0</Count><Status>Term or Database is not found</Status></ResultItem>
        (...)
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{GQuery}
\framesubtitle{Transforming to HTML using XSLT}
The XSLT stylesheet.
\lstinputlisting[language=xml,basicstyle=\tiny,breaklines=true,numbers=left]{gquery2html.xsl}
\end{frame}

\begin{frame}[fragile]
\frametitle{GQuery}
\framesubtitle{Transforming to HTML}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
 $ curl "http://eutils.ncbi.nlm.nih.gov/gquery?term=tyrannosaurus%20rex&retmode=xml" |\
 xsltproc gquery2html.xsl  -
\end{lstlisting}

\begin{lstlisting}[language=html,basicstyle=\tiny,breaklines=false]
<html>
  <body>
    <table>
      <caption>tyrannosaurus rex</caption>
      <tr>
        <th>Database</th>
        <th>Count</th>
        <th>Status</th>
      </tr>
      <tr>
        <td>
          <a href="http://www.ncbi.nlm.nih.gov/pubmed?cmd=search&amp;term=tyrannosaurus+rex">pubmed</a>
        </td>
        <td>41</td>
        <td>Ok</td>
      </tr>
      <tr>
        <td>
          <a href="http://www.ncbi.nlm.nih.gov/pubmed?cmd=search&amp;term=tyrannosaurus+rex">pmc</a>
        </td>
        <td>160</td>
        <td>Ok</td>
      </tr>
      <tr>
        <td>
          <a href="http://www.ncbi.nlm.nih.gov/pubmed?cmd=search&amp;term=tyrannosaurus+rex">mesh</a>
        </td>
        <td>15</td>
\end{lstlisting}
\end{frame}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% ESearch
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\hugeslide{ESearch}

\begin{frame}[fragile]
\frametitle{ESearch}
\begin{itemize}
\item Provides a list of UIDs matching a text query
\item Posts the results of a search on the History server
\item Downloads all UIDs from a dataset stored on the History server
\item Combines or limits UID datasets stored on the History server
\item Sorts sets of UIDs
\end{itemize}
\end{frame}


\begin{frame}[fragile]
\frametitle{ESearch}
\framesubtitle{Syntax}
Base URL
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi}
\end{frame}

\begin{frame}[fragile]
\frametitle{ESearch}
\framesubtitle{Searching for 'Mammuthus primigenius' }
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&term=%22Mammuthus%20primigenius%22%5BORGN%5D" |\
xmllint --format -
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<eSearchResult>
  <Count>684</Count>
  <RetMax>20</RetMax>
  <RetStart>0</RetStart>
  <IdList>
    <Id>507866428</Id>
    <Id>124056416</Id>
    <Id>383843869</Id>
    <Id>383843867</Id>
    <Id>383843865</Id>
    <Id>383843863</Id>
    <Id>383843861</Id>
    <Id>383843859</Id>
    <Id>383843857</Id>
    <Id>383843855</Id>
    <Id>383843853</Id>
    <Id>383843851</Id>
    <Id>383843849</Id>
    <Id>383843847</Id>
    <Id>383843845</Id>
    <Id>157367690</Id>
    <Id>157367676</Id>
    <Id>157367662</Id>
    <Id>157367648</Id>
    <Id>157367634</Id>
  </IdList>
  <TranslationSet>
    <Translation>
      <From>"Mammuthus primigenius"[ORGN]</From>
      <To>"Mammuthus primigenius"[Organism]</To>
    </Translation>
  </TranslationSet>
  <TranslationStack>
    <TermSet>
      <Term>"Mammuthus primigenius"[Organism]</Term>
      <Field>Organism</Field>
      <Count>684</Count>
      <Explode>Y</Explode>
    </TermSet>
    <OP>GROUP</OP>
  </TranslationStack>
  <QueryTranslation>"Mammuthus primigenius"[Organism]</QueryTranslation>
</eSearchResult>
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{ESearch}
\framesubtitle{the retmax parameter }
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&term=%22Mammuthus%20primigenius%22%5BORGN%5D&retmax=2" |\
xmllint --format -
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<eSearchResult>
  <Count>684</Count>
  <RetMax>2</RetMax>
  <RetStart>0</RetStart>
  <IdList>
    <Id>507866428</Id>
    <Id>124056416</Id>
  </IdList>
  <TranslationSet>
    <Translation>
      <From>"Mammuthus primigenius"[ORGN]</From>
      <To>"Mammuthus primigenius"[Organism]</To>
    </Translation>
  </TranslationSet>
  <TranslationStack>
    <TermSet>
      <Term>"Mammuthus primigenius"[Organism]</Term>
      <Field>Organism</Field>
      <Count>684</Count>
      <Explode>Y</Explode>
    </TermSet>
    <OP>GROUP</OP>
  </TranslationStack>
  <QueryTranslation>"Mammuthus primigenius"[Organism]</QueryTranslation>
</eSearchResult>
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{ESearch}
\framesubtitle{the retstart parameter }
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&term=%22Mammuthus%20primigenius%22%5BORGN%5D&retmax=3&retstart=100" |\
xmllint --format -
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<eSearchResult>
  <Count>684</Count>
  <RetMax>3</RetMax>
  <RetStart>100</RetStart>
  <IdList>
    <Id>300810656</Id>
    <Id>300810655</Id>
    <Id>300810654</Id>
  </IdList>
  <TranslationSet>
    <Translation>
      <From>"Mammuthus primigenius"[ORGN]</From>
      <To>"Mammuthus primigenius"[Organism]</To>
    </Translation>
  </TranslationSet>
  <TranslationStack>
    <TermSet>
      <Term>"Mammuthus primigenius"[Organism]</Term>
      <Field>Organism</Field>
      <Count>684</Count>
      <Explode>Y</Explode>
    </TermSet>
    <OP>GROUP</OP>
  </TranslationStack>
  <QueryTranslation>"Mammuthus primigenius"[Organism]</QueryTranslation>
</eSearchResult>
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{ESearch}
\framesubtitle{rettype=retcount  }
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=nucleotide&term=%22Mammuthus%20primigenius%22%5BORGN%5D&rettype=count" |\
xmllint --format -
\end{lstlisting}

\begin{lstlisting}[language=xml,breaklines=false]
<eSearchResult>
  <Count>684</Count>
</eSearchResult>
\end{lstlisting}
\end{frame}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% ESummary
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\hugeslide{ESummary}

\begin{frame}[fragile]
\frametitle{ESummary}
\framesubtitle{Syntax}
\begin{itemize}
\item Returns document summaries (DocSums) for a list of input UIDs
\item Returns DocSums for a set of UIDs stored on the Entrez History server
\end{itemize}
\end{frame}

\begin{frame}[fragile]
\frametitle{ESummary}
\framesubtitle{Syntax}
Base URL:
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=(DB)\&id=(TERM)}
\end{frame}


\begin{frame}[fragile]
\frametitle{ESummary}
\framesubtitle{Retrieve nucleotide gi=507866428}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
$ curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=nucleotide&id=507866428"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<eSummaryResult>
<DocSum>
	<Id>507866428</Id>
	<Item Name="Caption" Type="String">KC524742</Item>
	<Item Name="Title" Type="String">Mammuthus primigenius isolate CME2005/915 myoglobin (Mb) gene, partial cds</Item>
	<Item Name="Extra" Type="String">gi|507866428|gb|KC524742.1|[507866428]</Item>
	<Item Name="Gi" Type="Integer">507866428</Item>
	<Item Name="CreateDate" Type="String">2013/06/15</Item>
	<Item Name="UpdateDate" Type="String">2013/06/21</Item>
	<Item Name="Flags" Type="Integer">0</Item>
	<Item Name="TaxId" Type="Integer">37349</Item>
	<Item Name="Length" Type="Integer">9042</Item>
	<Item Name="Status" Type="String">live</Item>
	<Item Name="ReplacedBy" Type="String"></Item>
	<Item Name="Comment" Type="String"><![CDATA[  ]]></Item>
</DocSum>
</eSummaryResult>
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{ESummary}
\framesubtitle{Retrieve snp rs25}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=snp&id=25"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<eSummaryResult>
<DocSum>
	<Id>25</Id>
	<Item Name="SNP_ID" Type="Integer">25</Item>
	<Item Name="Organism" Type="String"></Item>
	<Item Name="ALLELE_ORIGIN" Type="String"></Item>
	<Item Name="GLOBAL_MAF" Type="String">0.4913</Item>
	<Item Name="GLOBAL_POPULATION" Type="String"></Item>
	<Item Name="GLOBAL_SAMPLESIZE" Type="Integer">0</Item>
	<Item Name="SUSPECTED" Type="String"></Item>
	<Item Name="CLINICAL_SIGNIFICANCE" Type="String"></Item>
	<Item Name="GENE" Type="String">THSD7A</Item>
	<Item Name="LOCUS_ID" Type="Integer">221981</Item>
	<Item Name="ACC" Type="String">NM_015204.2,NT_007819.17</Item>
	<Item Name="CHR" Type="String">7</Item>
	<Item Name="WEIGHT" Type="Integer">1</Item>
	<Item Name="HANDLE" Type="String">1000GENOMES,BGI,BL,BUSHMAN,COMPLETE_GENOMICS,CSHL-HAPMAP,GMI,ILLUMINA-UK,KWOK,PERLEGEN,SSMP,TISHKOFF</Item>
	<Item Name="FXN_CLASS" Type="String">intron-variant</Item>
	<Item Name="VALIDATED" Type="String">by-1000G,by-cluster,by-frequency,by-hapmap</Item>
	<Item Name="GTYPE" Type="String">true</Item>
	<Item Name="NONREF" Type="String">false</Item>
	<Item Name="DOCSUM" Type="String">HGVS=NC_000007.13:g.11584142T&gt;C,NG_027670.1:g.292683A&gt;G,NM_015204.2:c.1454-1398A&gt;G,NT_007819.17:g.11574142T&gt;C|SEQ=TCTGTGAGCTTCTGCATGCAATCCT[A/G]TGCAATTGGAATTTGATAGTCCTTT|GENE=THSD7A:221981</Item>
	<Item Name="HET" Type="Integer">50</Item>
	<Item Name="SRATE" Type="Integer">0</Item>
	<Item Name="TAX_ID" Type="Integer">9606</Item>
	<Item Name="CHRRPT" Type="String">25|2|0|1|1|1|7|NT_007819.17|11574141|11584142|THSD7A|0.499848|0.00872267| |51|1|1|36|138|0| | |T:2178:0.4913</Item>
	<Item Name="ORIG_BUILD" Type="Integer">36</Item>
	<Item Name="UPD_BUILD" Type="Integer">138</Item>
	<Item Name="CREATEDATE" Type="String">2000-09-19 17:02</Item>
	<Item Name="UPDATEDATE" Type="String">2013-06-21 14:17</Item>
	<Item Name="POP_CLASS" Type="String"></Item>
	<Item Name="METHOD_CLASS" Type="String">computed,hybridize,sequence,unknown</Item>
	<Item Name="SNP3D" Type="String"></Item>
	<Item Name="LINKOUT" Type="String">ILLUMINA-UK|http://www.illumina.com/HumanGenomeNA18507_000019106_NCBI36.1_chr7_11550667</Item>
	<Item Name="SS" Type="Integer">654151077</Item>
	<Item Name="LOCSNPID" Type="String">7_11584142</Item>
	<Item Name="ALLELE" Type="String">R</Item>
	<Item Name="SNP_CLASS" Type="String">snp</Item>
	<Item Name="CHRPOS" Type="String">7:11584142</Item>
	<Item Name="CONTIGPOS" Type="String">NT_007819.17:11574142</Item>
	<Item Name="TEXT" Type="String"></Item>
	<Item Name="LOOKUP" Type="String">325952</Item>
</DocSum>
</eSummaryResult>
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{ESummary}
\framesubtitle{Retrieve pubmed pmid=7939126}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
$ curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=pubmed&id=7939126"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<eSummaryResult>
<DocSum>
	<Id>7939126</Id>
	<Item Name="PubDate" Type="Date">1994 Apr</Item>
	<Item Name="EPubDate" Type="Date"></Item>
	<Item Name="Source" Type="String">Sleep</Item>
	<Item Name="AuthorList" Type="List">
		<Item Name="Author" Type="String">Broughton R</Item>
		<Item Name="Author" Type="String">Billings R</Item>
		<Item Name="Author" Type="String">Cartwright R</Item>
		<Item Name="Author" Type="String">Doucette D</Item>
		<Item Name="Author" Type="String">Edmeads J</Item>
		<Item Name="Author" Type="String">Edwardh M</Item>
		<Item Name="Author" Type="String">Ervin F</Item>
		<Item Name="Author" Type="String">Orchard B</Item>
		<Item Name="Author" Type="String">Hill R</Item>
		<Item Name="Author" Type="String">Turrell G</Item>
	</Item>
	<Item Name="LastAuthor" Type="String">Turrell G</Item>
	<Item Name="Title" Type="String">Homicidal somnambulism: a case report.</Item>
	<Item Name="Volume" Type="String">17</Item>
	<Item Name="Issue" Type="String">3</Item>
	<Item Name="Pages" Type="String">253-64</Item>
	<Item Name="LangList" Type="List">
		<Item Name="Lang" Type="String">English</Item>
	</Item>
	<Item Name="NlmUniqueID" Type="String">7809084</Item>
	<Item Name="ISSN" Type="String">0161-8105</Item>
	<Item Name="ESSN" Type="String">1550-9109</Item>
	<Item Name="PubTypeList" Type="List">
		<Item Name="PubType" Type="String">Journal Article</Item>
	</Item>
	<Item Name="RecordStatus" Type="String">PubMed - indexed for MEDLINE</Item>
	<Item Name="PubStatus" Type="String">ppublish</Item>
	<Item Name="ArticleIds" Type="List">
		<Item Name="pubmed" Type="String">7939126</Item>
		<Item Name="eid" Type="String">7939126</Item>
		<Item Name="rid" Type="String">7939126</Item>
	</Item>
	<Item Name="History" Type="List">
		<Item Name="pubmed" Type="Date">1994/04/01 00:00</Item>
		<Item Name="medline" Type="Date">1994/04/01 00:01</Item>
		<Item Name="entrez" Type="Date">1994/04/01 00:00</Item>
	</Item>
	<Item Name="References" Type="List"></Item>
	<Item Name="HasAbstract" Type="Integer">1</Item>
	<Item Name="PmcRefCount" Type="Integer">4</Item>
	<Item Name="FullJournalName" Type="String">Sleep</Item>
	<Item Name="ELocationID" Type="String"></Item>
	<Item Name="SO" Type="String">1994 Apr;17(3):253-64</Item>
</DocSum>
</eSummaryResult>
\end{lstlisting}
\end{frame}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% EFetch
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\hugeslide{EFetch}


\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Syntax}
Base URL:
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=(db)\&id=(ID)}
\end{frame}

\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Retrieve nucleotide gi=507866428 as ASN.1}

Default 
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=507866428}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
Seq-entry ::= set {
  class nuc-prot ,
  descr {
    source {
      genome genomic ,
      org {
        taxname "Mammuthus primigenius" ,
        common "woolly mammoth" ,
        db {
          {
            db "taxon" ,
            tag
              id 37349 } } ,
        orgname {
          name
            binomial {
              genus "Mammuthus" ,
              species "primigenius" } ,
          mod {
            {

\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Retrieve nucleotide gi=507866428 as Fasta}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=507866428&rettype=fasta}
\begin{lstlisting}[basicstyle=\tiny,breaklines=true]
>gi|507866428|gb|KC524742.1| Mammuthus primigenius isolate CME2005/915 myoglobin (Mb) gene, partial cds
GCACTTGCTTTTTTTGTCTTCTTCAGACCACGACATGGGACTCAGCGACGGGGAATGGGAGTTGGTGTTG
AAAACCTGGGGGAAAGTGGAGGCTGACATCCCGGGCCATGGGCTGGAAGTCTTCGTCAGGTAAAGGAAGA
AATCCTGTGGCCCCCATCACCCACCCCNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Retrieve nucleotide gi=507866428 as TinySeq}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=507866428&rettype=fasta&retmode=xml}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<?xml version="1.0"?>
 <!DOCTYPE TSeqSet PUBLIC "-//NCBI//NCBI TSeq/EN" 
 <TSeqSet>
<TSeq>
  <TSeq_seqtype value="nucleotide"/>
  <TSeq_gi>507866428</TSeq_gi>
  <TSeq_accver>KC524742.1</TSeq_accver>
  <TSeq_taxid>37349</TSeq_taxid>
  <TSeq_orgname>Mammuthus primigenius</TSeq_orgnam
  <TSeq_defline>Mammuthus primigenius isolate CME2
  <TSeq_length>9042</TSeq_length>
  <TSeq_sequence>GCACTTGCTTTTTTTGTCTTCTTCAGACCACGA
</TSeq>
</TSeqSet>
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Retrieve nucleotide gi=507866428 as Genbank-xml}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=507866428&retmode=xml}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<GBSeq>
  <GBSeq_locus>KC524742</GBSeq_locus>
  <GBSeq_length>9042</GBSeq_length>
  <GBSeq_strandedness>double</GBSeq_strandedness>
  <GBSeq_moltype>DNA</GBSeq_moltype>
  <GBSeq_topology>linear</GBSeq_topology>
  <GBSeq_division>MAM</GBSeq_division>
  <GBSeq_update-date>21-JUN-2013</GBSeq_update-date>
  <GBSeq_create-date>15-JUN-2013</GBSeq_create-date>
  <GBSeq_definition>Mammuthus primigenius isolate CME2005/915 myoglobin (Mb) gene, partial cds</GBSeq_definition>
  <GBSeq_primary-accession>KC524742</GBSeq_primary-accession>
  <GBSeq_accession-version>KC524742.1</GBSeq_accession-version>
  <GBSeq_other-seqids>
    <GBSeqid>gb|KC524742.1|</GBSeqid>
    <GBSeqid>gi|507866428</GBSeqid>
  </GBSeq_other-seqids>
  <GBSeq_source>Mammuthus primigenius (woolly mammoth)</GBSeq_source>
  <GBSeq_organism>Mammuthus primigenius</GBSeq_organism>
(...)
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Retrieve nucleotide gi=507866428 as Genbank}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=507866428&rettype=gb}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=false]
LOCUS       KC524742                9042 bp    DNA     linear   MAM 21-JUN-2013
DEFINITION  Mammuthus primigenius isolate CME2005/915 myoglobin (Mb) gene,
            partial cds.
ACCESSION   KC524742
VERSION     KC524742.1  GI:507866428
KEYWORDS    .
SOURCE      Mammuthus primigenius (woolly mammoth)
  ORGANISM  Mammuthus primigenius
            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
            Mammalia; Eutheria; Afrotheria; Proboscidea; Elephantidae;
            Mammuthus.
REFERENCE   1  (bases 1 to 9042)
  AUTHORS   Mirceta,S., Signore,A.V., Burns,J.M., Cossins,A.R., Campbell,K.L.
            and Berenbrink,M.
  TITLE     Evolution of mammalian diving capacity traced by myoglobin net
            surface charge
  JOURNAL   Science 340 (6138), 1234192 (2013)
   PUBMED   23766330
REFERENCE   2  (bases 1 to 9042)
  AUTHORS   Signore,A.V., Campbell,K.L. and Poinar,H.N.
  TITLE     Direct Submission
  JOURNAL   Submitted (09-JAN-2013) Biological Sciences, University of
            Manitoba, 50 Sifton Road, Winnipeg, Manitoba R3T2N2, Canada
COMMENT     ##Assembly-Data-START##
            Sequencing Technology :: Sanger dideoxy sequencing
            ##Assembly-Data-END##
FEATURES             Location/Qualifiers
     source          1..9042
                     /organism="Mammuthus primigenius"
                     /mol_type="genomic DNA"
                     /isolate="CME2005/915"
                     /db_xref="taxon:37349"
                     /tissue_type="bone"
     gene            <35..>9042
                     /gene="Mb"
     mRNA            join(<35..129,5627..5849,8979..>9042)
                     /gene="Mb"
                     /product="myoglobin"
     CDS             join(35..129,5627..5849,8979..>9042)

\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Efetch works with the ACCESSION NUMBERS}
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=KC524742&rettype=gb}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=false]
LOCUS       KC524742                9042 bp    DNA     linear   MAM 21-JUN-2013
DEFINITION  Mammuthus primigenius isolate CME2005/915 myoglobin (Mb) gene,
            partial cds.
ACCESSION   KC524742
VERSION     KC524742.1  GI:507866428
KEYWORDS    .
SOURCE      Mammuthus primigenius (woolly mammoth)
  ORGANISM  Mammuthus primigenius
            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
            Mammalia; Eutheria; Afrotheria; Proboscidea; Elephantidae;
            Mammuthus.
REFERENCE   1  (bases 1 to 9042)
  AUTHORS   Mirceta,S., Signore,A.V., Burns,J.M., Cossins,A.R., Campbell,K.L.
            and Berenbrink,M.
  TITLE     Evolution of mammalian diving capacity traced by myoglobin net
            surface charge
  JOURNAL   Science 340 (6138), 1234192 (2013)
   PUBMED   23766330
REFERENCE   2  (bases 1 to 9042)
  AUTHORS   Signore,A.V., Campbell,K.L. and Poinar,H.N.
  TITLE     Direct Submission
  JOURNAL   Submitted (09-JAN-2013) Biological Sciences, University of
            Manitoba, 50 Sifton Road, Winnipeg, Manitoba R3T2N2, Canada
COMMENT     ##Assembly-Data-START##
            Sequencing Technology :: Sanger dideoxy sequencing
            ##Assembly-Data-END##
FEATURES             Location/Qualifiers
     source          1..9042
                     /organism="Mammuthus primigenius"
                     /mol_type="genomic DNA"
                     /isolate="CME2005/915"
                     /db_xref="taxon:37349"
                     /tissue_type="bone"
     gene            <35..>9042
                     /gene="Mb"
     mRNA            join(<35..129,5627..5849,8979..>9042)
                     /gene="Mb"
                     /product="myoglobin"
     CDS             join(35..129,5627..5849,8979..>9042)

\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Using the WebEnv parameter.}
Web environment string returned from a previous ESearch, EPost or ELink call. When provided, ESearch will post the results of the search operation to this pre-existing WebEnv.
\end{frame}

\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Using the WebEnv parameter.}
Searching extinct species in the NCBI taxonomy ('extinct[PROP]') 
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
curl  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?!\color{red}{usehistory=y}!&db=taxonomy&term=extinct%5BPROP%5D"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false,escapechar=\%]
<eSearchResult>
  <Count>145</Count>
  <RetMax>20</RetMax>
  <RetStart>0</RetStart>
  <QueryKey>%\color{red}{1}%</QueryKey>
  <WebEnv>%\color{red}{NCID\_1\_75550312\_130.14.18.34\_9001\_1375948145\_325582538}%</WebEnv>
  <IdList>
    <Id>1225531</Id>
    <Id>1225530</Id>
    <Id>1211276</Id>
    <Id>1211275</Id>
    <Id>1027716</Id>
    <Id>948961</Id>
    <Id>943952</Id>
    <Id>867394</Id>
    <Id>867393</Id>
    <Id>748142</Id>
    <Id>748141</Id>
    <Id>741158</Id>
    <Id>703576</Id>
    <Id>703571</Id>
    <Id>703559</Id>
    <Id>693865</Id>
    <Id>686441</Id>
    <Id>665113</Id>
    <Id>659069</Id>
    <Id>656807</Id>
  </IdList>
  <TranslationSet/>
  <TranslationStack>
    <TermSet>
      <Term>extinct[PROP]</Term>
      <Field>PROP</Field>
      <Count>145</Count>
      <Explode>N</Explode>
    </TermSet>
    <OP>GROUP</OP>
  </TranslationStack>
  <QueryTranslation>extinct[PROP]</QueryTranslation>
</eSearchResult>
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{EFetch}
\framesubtitle{Using the WebEnv parameter.}

Fetch the extinct species in the NCBI taxonomy ('extinct[PROP]') using the WebEnv parameter.

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&!\color{red}{query\_key=1\&WebEnv=NCID\_1\_75550312\_130.14.18.34\_9001\_1375948145\_325582538}!&retmode=xml" 
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false,escapechar=\%]
<TaxaSet><Taxon>
    <TaxId>1225531</TaxId>
    <ScientificName>Equus ovodovi</ScientificName>
    <OtherNames>
        <Synonym>Equus (Sussemionus) ovodovi</Synonym>
        <Name>
            <ClassCDE>authority</ClassCDE>
            <DispName>Equus ovodovi Eisenmann &amp; Sergej, 2011</DispName>
        </Name>
    </OtherNames>
    <ParentTaxId>1225530</ParentTaxId>
    <Rank>species</Rank>
    <Division>Mammals</Division>
    <GeneticCode>
        <GCId>1</GCId>
        <GCName>Standard</GCName>
    </GeneticCode>
    <MitoGeneticCode>
    (....)
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{EPost}
\lstinputlisting[language=xml,basicstyle=\tiny,breaklines=true,numbers=left]{ncbi2epost.xsl}
\end{frame}

\begin{frame}[fragile]
\frametitle{EPost}
\framesubtitle{Post gi to epost}
Transforms the result of EFetch to a Bash script calling curl+Epost.
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&!\color{red}{query\_key=1\&WebEnv=NCID\_1\_75550312\_130.14.18.34\_9001\_1375948145\_325582538}!&retmode=xml" |\
!\color{red}{xsltproc --stringparam WebEnv NCID\_1\_75550312\_130.14.18.34\_9001\_1375948145\_325582538 ncbi2epost.xsl}! -
\end{lstlisting}

Output of XSLTproc
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=true,escapechar=\%]
curl -d @-  "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/epost.fcgi" <<EOF
db=taxonomy&WebEnv=NCID_1_75550312_130.14.18.34_9001_1375948145_325582538&id=1225531,1225530,1211276,1211275,1027716,948961,943952,867394,867393,748142,748141,741158,703576,703571,703559,693865,686441,665113,659069,656807,647691,647690,643746,643745,643744,643742,577682,572106,572105,572104,572099,572098,570943,570942,570941,551196,544298,523825,523824,523822,523821,523820,518692,518691,518689,475185,436495,436494,436493,436488,402889,399386,399178,386524,379504,363580,363579,363578,363571,339614,339612,339609,328612,314500,307641,304335,272462,268291,251263,251094,251093,239970,239969,237965,230980,230979,227166,227165,223567,222863,222862,216182,216181,201717,201716,192211,188536,187135,187134,187133,187132,187131,187118,184920,180214,173935,148923,147494,147466,147464,136416,136415,126594,126429,115942,107030,103864,94623,92649,92648,89252,89250,63631,63221,54568,54500,54497,54366,54365,48784,46906,39097,39053,39051,37349,37348,37185,27445,27444,20678,13266,13140,9619,9275,9274,9273,8818,8817,8815,8813,8812,8811,8810,8367,3409,
EOF
\end{lstlisting}

\end{frame}


\begin{frame}[fragile]
\frametitle{EPost}
\framesubtitle{Post gi to epost}
... the result is now piped into {\bf bash}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=taxonomy&!\color{red}{query\_key=1\&WebEnv=NCID\_1\_75550312\_130.14.18.34\_9001\_1375948145\_325582538}!&retmode=xml" |\
!\color{red}{xsltproc --stringparam WebEnv NCID\_1\_75550312\_130.14.18.34\_9001\_1375948145\_325582538 ncbi2epost.xsl}! - |\
!\color{red}{bash}
\end{lstlisting}
NCBI answer:
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=true,escapechar=\%]
<?xml version="1.0"?>
<!DOCTYPE ePostResult PUBLIC "-//NLM//DTD ePostResult, 11 May 2002//EN" "http://www.ncbi.nlm.nih.gov/entrez/query/DTD/ePost_020511.dtd">
<ePostResult>
	<QueryKey>2</QueryKey>
	<WebEnv>NCID_1_75550312_130.14.18.34_9001_1375948145_325582538</WebEnv>
</ePostResult>
\end{lstlisting}

\end{frame}

\begin{frame}[fragile]
\frametitle{EPost}
\framesubtitle{Searching in the WebEnv}
Search {\bf Homo Sapiens} in {\bf WebEnv} ?

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
curl -s "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=taxonomy&term=!\color{red}{Homo\%20Sapiens}!&usehistory=y&WebEnv=NCID_1_75550312_130.14.18.34_9001_1375948145_325582538&query_key=2"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=true,escapechar=\%]
<eSearchResult>
  <Count>0</Count>
  <RetMax>0</RetMax>
  <RetStart>0</RetStart>
  <QueryKey>8</QueryKey>
  <WebEnv>NCID_1_75550312_130.14.18.34_9001_1375948145_325582538</WebEnv>
  <IdList/>
  <TranslationSet/>
  <TranslationStack>
    <OP>GROUP</OP>
    <TermSet>
      <Term>homo sapiens[All Names]</Term>
      <Field>All Names</Field>
      <Count>1</Count>
      <Explode>N</Explode>
    </TermSet>
    <OP>AND</OP>
  </TranslationStack>
  <QueryTranslation>(#2) AND homo sapiens[All Names]</QueryTranslation>
</eSearchResult>
\end{lstlisting}

\end{frame}

\begin{frame}[fragile]
\frametitle{EPost}
\framesubtitle{Searching in the WebEnv}
Search {\bf Tyranosaurus} in {\bf WebEnv} ?
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl -s "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=taxonomy&term=!\color{red}{Tyrannosaurus}!&usehistory=y&WebEnv=NCID_1_75550312_130.14.18.34_9001_1375948145_325582538&query_key=2"
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=true,escapechar=\%]
<eSearchResult>
  <Count>1</Count>
  <RetMax>1</RetMax>
  <RetStart>0</RetStart>
  <QueryKey>9</QueryKey>
  <WebEnv>NCID_1_75550312_130.14.18.34_9001_1375948145_325582538</WebEnv>
  <IdList>
    <Id>436494</Id>
  </IdList>
  <TranslationSet/>
  <TranslationStack>
    <OP>GROUP</OP>
    <TermSet>
      <Term>Tyrannosaurus[All Names]</Term>
      <Field>All Names</Field>
      <Count>1</Count>
      <Explode>N</Explode>
    </TermSet>
    <OP>AND</OP>
  </TranslationStack>
  <QueryTranslation>(#2) AND Tyrannosaurus[All Names]</QueryTranslation>
</eSearchResult>
\end{lstlisting}
\end{frame}



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% ELINK
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\hugeslide{Elink}

\begin{frame}[fragile]
\frametitle{Elink}
\begin{itemize}
\item Returns UIDs linked to an input set of UIDs in either the same or a different Entrez database
\item Returns UIDs linked to other UIDs in the same Entrez database that match an Entrez query
\item Checks for the existence of Entrez links for a set of UIDs within the same database
\item Lists the available links for a UID
\item Lists LinkOut URLs and attributes for a set of UIDs
\item Lists hyperlinks to primary LinkOut providers for a set of UIDs
\item Creates hyperlinks to the primary LinkOut provider for a single UID
\end{itemize}
\end{frame}

\begin{frame}[fragile]
\frametitle{Elink}
Base URL:
\small
\url{http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi}
\end{frame}


\begin{frame}[fragile]
\frametitle{ELink}
\framesubtitle{Searching the pubmed records associated to sequence gi:507866428}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
http://eutils.ncbi.nlm.nih.gov/entrez/eutils/elink.fcgi?dbfrom=nucleotide&db=pubmed&id=507866428&cmd=neighbor_score
\end{lstlisting}

\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=true,escapechar=\%]
<eLinkResult>
  <LinkSet>
    <DbFrom>nuccore</DbFrom>
    <IdList>
      <Id>507866428</Id>
    </IdList>
    <LinkSetDb>
      <DbTo>pubmed</DbTo>
      <LinkName>nuccore_pubmed</LinkName>
      <Link>
        <Id>23766330</Id>
        <Score>0</Score>
      </Link>
    </LinkSetDb>
  </LinkSet>
</eLinkResult>

\end{lstlisting}

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl -s "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&id=23766330&rettype=medline&retmode=text"
PMID- 23766330
TI  - Evolution of mammalian diving capacity traced by myoglobin net surface charge.
PG  - 1234192
LID - 10.1126/science.1234192 [doi]
\end{lstlisting}
\end{frame}


\hugeslide{Transformations}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% SVG
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to SVG}

Using the stylesheet \url{https://github.com/lindenb/xslt-sandbox/blob/master/stylesheets/bio/ncbi/gb2svg.xsl}

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
xsltproc <(curl  "https://raw.github.com/lindenb/xslt-sandbox/master/stylesheets/bio/ncbi/gb2svg.xsl") \
	"http://www.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=14971102&retmode=xml&rettype=gbc"
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to SVG}
\lstinputlisting[language=xml,basicstyle=\tiny,breaklines=true,numbers=left]{jeter01.svg}
\end{frame}

\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to SVG}
\begin{center}
\includegraphics[scale=0.4]{jeter01svg.png}
\end{center}
\end{frame}


\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to R}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl -s "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=Tyrannosaurus&usehistory=true" | xmllint --format -
$ curl -s "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&usehistory=true&WebEnv=NCID_1_52434791_130.14.22.215_9001_1375957034_1619786167&query_key=1&retmode=xml"
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to R}
\lstinputlisting[language=xml,basicstyle=\tiny,breaklines=true,numbers=left]{pubmed2rstats.xsl}
\end{frame}

\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to R}

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&usehistory=true&WebEnv=NCID_1_52434791_130.14.22.215_9001_1375957034_1619786167&query_key=1&retmode=xml" |\
xsltproc pubmed2rstats.xsl -
\end{lstlisting}

\begin{lstlisting}[language=R,basicstyle=\tiny,breaklines=true,escapechar=\!]
date2count <- list()

date2count[["2013"]] <- ifelse(is.null(date2count[["2013"]]),1,1+date2count[["2013"]])
date2count[["2012"]] <- ifelse(is.null(date2count[["2012"]]),1,1+date2count[["2012"]])
date2count[["2012"]] <- ifelse(is.null(date2count[["2012"]]),1,1+date2count[["2012"]])
date2count[["2011"]] <- ifelse(is.null(date2count[["2011"]]),1,1+date2count[["2011"]])
date2count[["2011"]] <- ifelse(is.null(date2count[["2011"]]),1,1+date2count[["2011"]])
(..)
df <- data.frame(
	Year=as.integer(names(date2count)),
	Count=unlist(date2count)
	)
png('jeterpubmed.png')
plot(df)
title('pubmed: count(articles)=f(year)')
dev.off()
\end{lstlisting}

\end{frame}



\begin{frame}[fragile]
\frametitle{Efetch}
\framesubtitle{Transforming to R}

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true,escapechar=\!]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&usehistory=true&WebEnv=NCID_1_52434791_130.14.22.215_9001_1375957034_1619786167&query_key=1&retmode=xml" |\
xsltproc pubmed2rstats.xsl - |\
R --no-save
\end{lstlisting}

\begin{center}
\includegraphics[scale=0.3]{jeterpubmed.png}
\end{center}
\end{frame}

\hugeslide{Generating a JAVA parser}


\begin{frame}[fragile]
\frametitle{Using the XML schema}
\framesubtitle{XML Schema for dbSNP}
\url{ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_3.3.xsd}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
<?xml version="1.0" encoding="UTF-8"?>
<xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns="http://www.ncbi.nlm.nih.gov/SNP/docsum" targetNamespace="http://www.ncbi.nlm.nih.gov/SNP/docsum" el
ementFormDefault="qualified" attributeFormDefault="unqualified">
  <xsd:element name="ExchangeSet">
    <xsd:annotation>
      <xsd:documentation>Set of dbSNP refSNP docsums, version 3.2</xsd:documentation>
    </xsd:annotation>
    <xsd:complexType>
      <xsd:sequence>
        <xsd:element name="SourceDatabase" minOccurs="0">
          <xsd:complexType>
            <xsd:attribute name="taxId" type="xsd:int" use="required">
              <xsd:annotation>
                <xsd:documentation>NCBI taxonomy ID for variation</xsd:documentation>
              </xsd:annotation>
            </xsd:attribute>
            <xsd:attribute name="organism" type="xsd:string" use="required">
              <xsd:annotation>
                <xsd:documentation>common name for species used as part of database name.</xsd:documentation>
              </xsd:annotation>
            </xsd:attribute>
            <xsd:attribute name="dbSnpOrgAbbr" type="xsd:string">
              <xsd:annotation>
                <xsd:documentation>organism abbreviation used in dbSNP. </xsd:documentation>
              </xsd:annotation>
            </xsd:attribute>
            <xsd:attribute name="gpipeOrgAbbr" type="xsd:string">
              <xsd:annotation>
                <xsd:documentation>organism abbreviation used within NCBI genome pipeline data dumps.</xsd:documentation>
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{Using the XML schema}
\framesubtitle{Compiling the XML Schema for dbSNP with XJC}

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=false]
$ xjc -d . "ftp://ftp.ncbi.nlm.nih.gov/snp/specs/docsum_3.3.xsd"
parsing a schema...
compiling a schema...
gov/nih/nlm/ncbi/snp/docsum/Assay.java
gov/nih/nlm/ncbi/snp/docsum/Assembly.java
gov/nih/nlm/ncbi/snp/docsum/BaseURL.java
gov/nih/nlm/ncbi/snp/docsum/Component.java
gov/nih/nlm/ncbi/snp/docsum/ExchangeSet.java
gov/nih/nlm/ncbi/snp/docsum/FxnSet.java
gov/nih/nlm/ncbi/snp/docsum/MapLoc.java
gov/nih/nlm/ncbi/snp/docsum/ObjectFactory.java
gov/nih/nlm/ncbi/snp/docsum/PrimarySequence.java
gov/nih/nlm/ncbi/snp/docsum/Rs.java
gov/nih/nlm/ncbi/snp/docsum/RsLinkout.java
gov/nih/nlm/ncbi/snp/docsum/RsStruct.java
gov/nih/nlm/ncbi/snp/docsum/Ss.java
gov/nih/nlm/ncbi/snp/docsum/package-info.java
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Using the XML schema}
\framesubtitle{Compiling the XML Schema for dbSNP with XJC}
Search the non-genomic rs\# in dbSNP.
\lstinputlisting[language=java,basicstyle=\tiny,breaklines=true,numbers=left]{ParseDbSnp.java}
\end{frame}

\begin{frame}[fragile]
\frametitle{Using the XML schema}
\framesubtitle{Compiling the XML Schema for dbSNP with XJC}
compile...
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=false]
$ javac ParseDbSnp.java gov/nih/nlm/ncbi/snp/docsum/*.java
\end{lstlisting}
and run...
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=false]
$ curl -s "ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606/XML/ds_ch1.xml.gz" |\
 gunzip -c |\
 java ParseDbSnp
rs701 cDNA
rs860 cDNA
rs861 cDNA
rs862 cDNA
rs863 cDNA
rs864 cDNA
rs865 cDNA
rs866 cDNA
rs877 cDNA
rs878 cDNA
rs879 cDNA
rs880 cDNA
rs882 cDNA
rs883 cDNA
rs884 cDNA
rs885 cDNA
rs886 cDNA
rs913 cDNA
rs945 cDNA
rs946 cDNA
(...)
\end{lstlisting}
\end{frame}


\hugeslide{NCBI EBot}

\begin{frame}[fragile]
\frametitle{NCBI EBot}
\framesubtitle{URL}
\url{http://www.ncbi.nlm.nih.gov/Class/PowerTools/eutils/ebot/ebot.cgi}
\end{frame}

\begin{frame}[fragile]
\frametitle{NCBI EBot}
\framesubtitle{Sample output}
\begin{lstlisting}[language=perl,basicstyle=\tiny,breaklines=true]
#!/usr/bin/perl 
(...)
#                            PUBLIC DOMAIN NOTICE
#               National Center for Biotechnology Information
use LWP::Simple;
use LWP::UserAgent;
use Net::FTP;


my $delay = 0;
my $maxdelay = 3;
my $base = "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/";


$params{email} = "nobody@nowhere.com";
$params{db} = "nuccore";
$params{tool} = "ebot";
$params{term} = "Mammuthus+primigenius[ORGN]";
%params = esearch(%params);

$params{retmode} = "xml";
$params{outfile} = "result.xml";
$params{rettype} = "native";
efetch_batch(%params);
\end{lstlisting}
\end{frame}

\hugeslide{BLAST}

\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Downloading}
Standalone tools are available at \url{ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/}
\begin{lstlisting}[language=bash]
#add BLAST to your path
export PATH=${PATH}:/path/to/ncbi-blast-2.2.28+/bin
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Download a sample}
apis mellifera proteins
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
curl -o protein.fa.gz \
   "ftp://ftp.ncbi.nih.gov/genomes/Apis_mellifera/protein/protein.fa.gz"
gunzip protein.fa.gz
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Create a Blast database with makeblastdb}
Getting help...
\begin{lstlisting}[language=bash]
$ makeblastdb -help
(...)
 -dbtype <String, `nucl', `prot'>
   Molecule type of target db
 -in <File_In>
   Input file/database name
   Default = `-'
 -input_type <String, `asn1_bin', `asn1_txt', `blastdb', `fasta'>
   Type of the data specified in input_file
   Default = `fasta'
(..)
\end{lstlisting}
\end{frame}


\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Create a Blast database with makeblastdb}
Create the BLAST database:
\begin{lstlisting}[language=bash]
$ makeblastdb -in protein.fa -dbtype prot

Building a new DB, current time: 09/02/2013 18:29:38
New DB name:   protein.fa
New DB title:  protein.fa
Sequence type: Protein
Keep Linkouts: T
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 10570 sequences in 1.84458 seconds.
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Query a Blast database with blastp}
Get help:
\begin{lstlisting}[language=bash]
$ blastp -help
(...)
 -query <File_In>
   Input file name
   Default = `-'
 -db <String>
   BLAST database name
(...)
\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Blast human EIF4G1 gi:187956781 }
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&rettype=fasta&id=187956781" |\
 blastp -db protein.fa
\end{lstlisting}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=false]
Query= gi|187956781|gb|AAI40897.1| EIF4G1 protein [Homo sapiens]
(...)
                                                                      Score     E
Sequences producing significant alignments:                          (Bits)  Value

  gi|328782175|ref|XP_394628.4| PREDICTED: eukaryotic translation...   189    4e-49
  gi|328779480|ref|XP_003249661.1| PREDICTED: hypothetical protei...  38.1    0.017
  gi|110762568|ref|XP_001121713.1| PREDICTED: hypothetical protei...  38.1    0.018

(...)
> gi|328782175|ref|XP_394628.4| PREDICTED: eukaryotic translation 
initiation factor 4 gamma 2-like [Apis mellifera]
Length=899

 Score =  189 bits (479),  Expect = 4e-49, Method: Compositional matrix adjust.
 Identities = 115/319 (36%), Positives = 175/319 (55%), Gaps = 39/319 (12%)

Query  717   KEPRKIIATVLMTEDIKLNKAEKAWKPSS--KRTAADKDRGEEDADGSKTQDLFRRVRSI  774
             ++P +    +++ +DI+    E+ W P S  +R A   +        S+   +FR+VR I
Sbjct  22    RKPSETTVGLVIKDDIRSLSTEQRWIPPSTLRRDALTPE--------SRNNFIFRKVRGI  73

Query  775   LNKLTPQMFQQLMKQVTQLAIDTEERLKGVIDLIFEKAISEPNFSVAYANMCRCL-----  829
             LNKLTP+ F +L   +  + ++++  LKGVI LIFEKA+ EP +S  YA +C+ L     
Sbjct  74    LNKLTPEKFAKLSNDLLNVELNSDVILKGVIFLIFEKALDEPKYSSMYAQLCKRLSDEAA  133

Query  830   -MALKVPTTEKPTVTVNFRKLLLNRCQKEFEKDKDDDEVFEKKQKEMDEAATAEERGRLK  888
                 K    E       F  LLL++C+ EFE      E FE +    DE    EE     
Sbjct  134   NFEPKKALIESQKGQSTFTFLLLSKCRDEFENRSKASEAFENQ----DELGPEEE-----  184

\end{lstlisting}
\end{frame}

\begin{frame}[fragile]
\frametitle{Standalone Blast}
\framesubtitle{Blast human EIF4G1 gi:187956781 , ouput XML}
\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
$ curl "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=protein&rettype=fasta&id=187956781" |\
 blastp -db protein.fa  -outfmt  5
\end{lstlisting}
\begin{lstlisting}[language=xml,basicstyle=\tiny,breaklines=false]
(...)
 <Hit_hsps>
    <Hsp>
      <Hsp_num>1</Hsp_num>
      <Hsp_bit-score>189.119</Hsp_bit-score>
      <Hsp_score>479</Hsp_score>
      <Hsp_evalue>3.78314e-49</Hsp_evalue>
      <Hsp_query-from>717</Hsp_query-from>
      <Hsp_query-to>1017</Hsp_query-to>
      <Hsp_hit-from>22</Hsp_hit-from>
      <Hsp_hit-to>319</Hsp_hit-to>
      <Hsp_query-frame>0</Hsp_query-frame>
      <Hsp_hit-frame>0</Hsp_hit-frame>
      <Hsp_identity>115</Hsp_identity>
      <Hsp_positive>175</Hsp_positive>
      <Hsp_gaps>39</Hsp_gaps>
      <Hsp_align-len>319</Hsp_align-len>
      <Hsp_qseq>KEPRKIIATVLMTEDIKLNKAEKAWKPSS--KRTAADKDRGEEDADGSKTQDLFRRVRSILNKLTPQMFQQLMKQVTQLAIDTEERLKGVIDLIFEKAISEPNFSVAYANMCRCL------MALKVPTTEKPTVTVNFRKLLLNRCQKEFEKDKDDDEVFEKKQKEMDEAATAEERGRLKEELEEARD
IARRRSLGNIKFIGELFKLKMLTEAIMHDCVVKLL--------KNHDEESLECLCRLLTTIGKDLDFEKAKPRMDQYFNQMEKIIKEKKTSSRIRFMLQDVLDLRGSNWVPRRG--DQGPKTIDQIHKEAE</Hsp_qseq>
      <Hsp_hseq>RKPSETTVGLVIKDDIRSLSTEQRWIPPSTLRRDALTPE--------SRNNFIFRKVRGILNKLTPEKFAKLSNDLLNVELNSDVILKGVIFLIFEKALDEPKYSSMYAQLCKRLSDEAANFEPKKALIESQKGQSTFTFLLLSKCRDEFENRSKASEAFENQ----DELGPEEE---------ERRQ
VAKRKMLGNIKFIGELGKLGIVSETILHRCILQLLEKKRRRRSRGDTAEDIECLCQIMRTCGRILDSDKGRGLMDQYFKRMNSLAESRDLPLRIKFMLRDVIELRRDGWVPRKATSTEGPMPINQIRNDNE</Hsp_hseq>
      <Hsp_midline>++P +    +++ +DI+    E+ W P S  +R A   +        S+   +FR+VR ILNKLTP+ F +L   +  + ++++  LKGVI LIFEKA+ EP +S  YA +C+ L         K    E       F  LLL++C+ EFE      E FE +    DE    EE         E
 R +A+R+ LGNIKFIGEL KL +++E I+H C+++LL        +    E +ECLC+++ T G+ LD +K +  MDQYF +M  + + +    RI+FML+DV++LR   WVPR+    +GP  I+QI  + E</Hsp_midline>
    </Hsp>
(...)
\end{lstlisting}
\end{frame}


\hugeslide{NCBI URL-API Blast}

\begin{frame}[fragile]
\frametitle{NCBI URL-API Blast}

http://www.ncbi.nlm.nih.gov/blast/Doc/urlapi.html

\begin{lstlisting}[language=bash,basicstyle=\tiny,breaklines=true]
$ curl "http://www.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Put&QUERY=PAERLMERKADIE&DATABASE=nr&PROGRAM=blastp&FILTER=L&HITLIST_SZE=500"
\end{lstlisting}
\begin{lstlisting}[language=html,breaklines=false]
(...)
 
<!--QBlastInfoBegin
    RID = 1NRYGX9K014
    RTOE = 29
QBlastInfoEnd
-->

(...)
\end{lstlisting}
\end{frame}


\hugeslide{The End}
\end{document}

